### PR TITLE
Upstream several Intel extensions

### PR DESCRIPTION
**Pull request by MrSidims**

### CODE REVIEW

**Comment by MrSidims** [comment]
@bashbaug @mkinsner could you please take a look?

**Comment by MrSidims** [comment]
@mkinsner  @bashbaug could you please take a look and possibly merge? It looks like misalignment between SPIR-V headers starts bringing and attention: https://github.com/intel/llvm/issues/2889

File: include/spirv/unified1/spirv.core.grammar.json
Comment (dneto0) [inline include/spirv/unified1/spirv.core.grammar.json:5113]: I just wanted to check that this is an intentional change.

File: include/spirv/unified1/spirv.core.grammar.json
Comment (dneto0) [inline include/spirv/unified1/spirv.core.grammar.json:9338]: Adding a new operand kind like this will break the SPIRV-Tools build.
I plan to prepare a patch to SPIRV-Tools to absorb this change.

This is the kind of error I get when rebasing this change against master and rebuilding SPIRV-Tools:


In file included from ...spirv-tools/source/operand.cpp:36:
third_party/spirv-tools/operand.kinds-unified1.inc:703:172: error: ‘SPV_OPERAND_TYPE_FPDENORM_MODE’ was not declared in this scope; did you mean ‘SPV_OPERAND_TYPE_MEMORY_MODEL’?
  703 |   {"FunctionDenormModeINTEL", 5823, 1, pygen_variable_caps_FunctionFloatControlINTEL, 1, pygen_variable_exts_SPV_INTEL_float_controls2, {SPV_OPERAND_TYPE_LITERAL_INTEGER, SPV_OPERAND_TYPE_FPDENORM_MODE}, 0xffffffffu, 0xffffffffu},
      |                                                                                                                                                                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                                                                                                                                            SPV_OPERAND_TYPE_MEMORY_MODEL
third_party/spirv-tools/operand.kinds-unified1.inc:724:179: error: ‘SPV_OPERAND_TYPE_FPOPERATION_MODE’ was not declared in this scope; did you mean ‘SPV_OPERAND_TYPE_EXECUTION_MODE’?
  724 |   {"FunctionFloatingPointModeINTEL", 6080, 1, pygen_variable_caps_FunctionFloatControlINTEL, 1, pygen_variable_exts_SPV_INTEL_float_controls2, {SPV_OPERAND_TYPE_LITERAL_INTEGER, SPV_OPERAND_TYPE_FPOPERATION_MODE}, 0xffffffffu, 0xffffffffu},
      |                                                                                                                                                                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                                                                                                                                                   SPV_OPERAND_TYPE_EXECUTION_MODE
third_party/spirv-tools/operand.kinds-unified1.inc:1229:4: error: ‘SPV_OPERAND_TYPE_FPDENORM_MODE’ was not declared in this scope; did you mean ‘SPV_OPERAND_TYPE_MEMORY_MODEL’?
 1229 |   {SPV_OPERAND_TYPE_FPDENORM_MODE, ARRAY_SIZE(pygen_variable_FPDenormModeEntries), pygen_variable_FPDenormModeEntries},
      |    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |    SPV_OPERAND_TYPE_MEMORY_MODEL
third_party/spirv-tools/operand.kinds-unified1.inc:1230:4: error: ‘SPV_OPERAND_TYPE_FPOPERATION_MODE’ was not declared in this scope; did you mean ‘SPV_OPERAND_TYPE_EXECUTION_MODE’?
 1230 |   {SPV_OPERAND_TYPE_FPOPERATION_MODE, ARRAY_SIZE(pygen_variable_FPOperationModeEntries), pygen_variable_FPOperationModeEntries},
      |    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |    SPV_OPERAND_TYPE_EXECUTION_MODE

**Review by dneto0 — COMMENTED** [review]
So far this change looks ok.
For one thing, I verified that all the enum values are placed in the right order.

I'll need to prepare a patch for SPIRV-Tools to support the new enumerant kinds FPDenorm_Mode and FPOperation_Mode

**Comment by dneto0** [comment]
Please rebase this change against master, to keep it fresh.

File: include/spirv/unified1/spirv.core.grammar.json
Comment (dneto0) [inline include/spirv/unified1/spirv.core.grammar.json:9342]: I think it is preferable to add:

     "capabilities" : [ "FunctionFloatControlINTEL" ],
     "version" : "None"

File: include/spirv/unified1/spirv.core.grammar.json
Comment (dneto0) [inline include/spirv/unified1/spirv.core.grammar.json:9355]: Similarly, I think it's preferable to add:

     "capabilities" : [ "FunctionFloatControlINTEL" ],
     "version" : "None"

File: include/spirv/unified1/spirv.core.grammar.json
Comment (dneto0) [reply include/spirv/unified1/spirv.core.grammar.json:9338]: I have a pending patch for SPIRV-Tools.  https://github.com/KhronosGroup/SPIRV-Tools/pull/4116  
That will have to land before this patch (to prevent the SPIRV-Tools build from breaking).

**Comment by raunraun** [comment]
Will merge when SPIRV-Tiils#4116 is given the OK by the bots.

File: include/spirv/unified1/spirv.core.grammar.json
Comment (MrSidims) [reply include/spirv/unified1/spirv.core.grammar.json:5113]: Yes, it is. This Op code was renamed with a new revision of the appropriate spec.

File: include/spirv/unified1/spirv.core.grammar.json
Comment (MrSidims) [reply include/spirv/unified1/spirv.core.grammar.json:9338]: Yeah, it also required changes in the headers builder itself.

File: include/spirv/unified1/spirv.core.grammar.json
Comment (MrSidims) [reply include/spirv/unified1/spirv.core.grammar.json]: Thanks! It wasn't mentioned in the [spec](https://github.com/intel/llvm/blob/39fa9b0cbfbae88327118990a05c5b387b56d2ef/sycl/doc/extensions/SPIRV/SPV_INTEL_float_controls2.asciidoc), but I guess you are right, I'll open a spec issue.

**Comment by MrSidims** [comment]
Shall I squash commits before merging to eliminate the last "Apply suggestions" commit?

File: include/spirv/unified1/spirv.core.grammar.json
Comment (dneto0) [reply include/spirv/unified1/spirv.core.grammar.json:9337]: I've landed the SPIRV-Tools change, so we're clear to land this one, from my perspective.

File: include/spirv/unified1/spirv.core.grammar.json
Comment (dneto0) [reply include/spirv/unified1/spirv.core.grammar.json]: Yes, for everything that isn't a capability.
The capability should keep it.

**Comment by dneto0** [comment]
> Shall I squash commits before merging to eliminate the last "Apply suggestions" commit?

I don't think it's necessary to squash.

**Review by dneto0 — APPROVED** [review]
Looks good to me.
I've checked this with the corresponding SPIRV-Tools change, and the combination works together.
Thanks!

**Comment by MrSidims** [comment]
> Looks good to me.
> I've checked this with the corresponding SPIRV-Tools change, and the combination works together.
> Thanks!

Good news! Thanks for making the things going!

**Pull request merged**
